### ISSUE TITLE
Nest testing tools: Undefined method prepareRequest()

### ISSUE THREAD

**Issue by weaverryan** [api-platform/core#2982]
Hi friends!

The new testing [Client](https://github.com/api-platform/core/blob/master/src/Bridge/Symfony/Bundle/Test/Client.php) class implements `Symfony\Contracts\HttpClient\HttpClientInterface`, which is cool! But, it also implicitly relies on a concrete method on `HttpClient` itself - the `prepareRequest()` method: https://github.com/api-platform/core/blob/92f655c84e9d03f38ee2986906fba11ce529bc2f/src/Bridge/Symfony/Bundle/Test/Client.php#L79

If you don't have http-client installed, you'll get a cryptic error about calling an undefined function.

**Edit** Oddly, I get an "undefined method" on the first run... and then a "Trait 'Symfony\Component\HttpClient\HttpClientTrait' not found" on the second... which makes more sense...

That function specifically lives in [HttpClientTrait](https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/HttpClient/HttpClientTrait.php)

In other words, as it's written now, `symfony/http-client` is a hard dependency on the new `Client` class. We could make this more user-friendly by checking for the presence of that trait before using it. That's not possible inside of `Client` itself (you can't conditionally `use` a trait), so... we could do it inside of `ApiTestCase::createClient()` instead... which is how a user would "usually" create the `Client`.

... or we could require `symfony/http-client` as a dependency of api-platform/core... though afaik, it's only needed in this one spot currently...

Thoughts?

**Comment by teohhanhui** [comment]
But why would the `test.api_platform.client` service exist, if the instantiated class `ApiPlatform\Core\Bridge\Symfony\Bundle\Test\Client` has incomplete definition? It sounds like something that should be fixed in Symfony DI?

**Comment by dunglas** [comment]
I agree that we should throw an error if the deps is missing!

**Comment by dunglas** [comment]
@teohhanhui it's not a bug in SF: we always create this service in test mode: https://github.com/api-platform/core/blob/master/src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php#L131-L133

We could indeed not create it if the HttpClient component is not installed, but I think that @weaverryan's proposal is more user friendly (and it has no performance drawback as this service is only registered in test mode).

**Comment by teohhanhui** [comment]
What I mean is, if we try to instantiate our `Client` class, it should already fail, as @weaverryan has pointed out:

```
Trait 'Symfony\Component\HttpClient\HttpClientTrait' not found
```

So indeed, it has nothing to do with Symfony DI, as it's not something that could be detected during container compilation.

**Comment by teohhanhui** [comment]
Perhaps we could throw an exception in the `Client` constructor?

**Comment by dunglas** [comment]
@teohhanhui it will not work: https://3v4l.org/pmIIr

**Comment by teohhanhui** [comment]
Yup. So actually I'm not sure we should do anything here. Since PHP already throws a fatal error because of the missing trait. It should be pretty obvious what the user needs to do.

**Comment by dunglas** [comment]
Having an error telling `Run "composer require symfony/http-client" to use the test client", as we already do in some cases (access control IIRC) is way better.

**Comment by weaverryan** [comment]
> Having an error telling Run "composer require symfony/http-client" to use the test client", as we already do in some cases (access control IIRC) is way better

Yep, that's basically what I was thinking. It throws an error currently, just not a great error :)
